src/libopensc/card-isoApplet.c

```diff
@@ -1195,21 +1235,24 @@ isoApplet_compute_signature(struct sc_card *card,
 		u8* p = NULL;
 		size_t len = (drvdata->sec_env_ec_field_length + 7) / 8 * 2;
 
-		if (len > outlen)
+		if (len > seqlen)
 			LOG_FUNC_RETURN(ctx, SC_ERROR_BUFFER_TOO_SMALL);
 
 		p = calloc(1,len);
 		if (!p)
 			LOG_FUNC_RETURN(ctx, SC_ERROR_OUT_OF_MEMORY);
 
-		r = sc_asn1_sig_value_sequence_to_rs(ctx, out, r, p, len);
+		r = sc_asn1_sig_value_sequence_to_rs(ctx, seqbuf, r, p, len);
 		if (!r)   {
-			memcpy(out, p, len);
+			memcpy(seqbuf, p, len);
 			r = len;
 		}
 
 		free(p);
 	}
+	if ((size_t) r > outlen)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_BUFFER_TOO_SMALL);
+	memcpy(out, seqbuf, r);
```

Comment:
I am wondering if we need these two memcpy. It will probably not be large overhead, but copying data back and forth might be at least confusing and error prone.

Reply:
Since we added an additional branch for RSA-PSS I refactored the code to handle each case (RSA, RSA-PSS and ECDSA) in its own branch. We only need seqbuf in the ECDSA branch and now we do not need additional memcpy's.